plugin/trino-hive/src/test/java/io/trino/plugin/hive/metastore/thrift/TestingThriftHttpMetastoreServer.java

```diff
@@ -0,0 +1,179 @@
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.apache.thrift.TException;
+import org.apache.thrift.TProcessor;
+import org.apache.thrift.protocol.TBinaryProtocol;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.protocol.TProtocolFactory;
+import org.apache.thrift.transport.TIOStreamTransport;
+import org.apache.thrift.transport.TTransport;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URI;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static com.google.common.reflect.Reflection.newProxy;
+
+public class TestingThriftHttpMetastoreServer
+        implements Closeable
+{
+    private final TestingThriftHttpServlet thriftHttpServlet;
+    private final LifeCycleManager lifeCycleManager;
+    private final URI baseUri;
+
+    public TestingThriftHttpMetastoreServer(ThriftMetastore delegate, Consumer<HttpServletRequest> requestInterceptor)
+    {
+        ThriftHiveMetastore.Iface mockThriftHandler = proxyHandler(delegate, ThriftHiveMetastore.Iface.class);
+        TProcessor processor = new ThriftHiveMetastore.Processor<>(mockThriftHandler);
+        thriftHttpServlet = new TestingThriftHttpServlet(processor, new TBinaryProtocol.Factory(), requestInterceptor);
+        Bootstrap app = new Bootstrap(
+                new TestingNodeModule(),
+                new TestingHttpServerModule(),
+                binder -> {
+                    binder.bind(new TypeLiteral<Map<String, String>>() {}).annotatedWith(TheServlet.class).toInstance(ImmutableMap.of());
+                    binder.bind(Servlet.class).annotatedWith(TheServlet.class).toInstance(thriftHttpServlet);
+                });
+
+        Injector injector = app
+                .doNotInitializeLogging()
+                .initialize();
+
+        lifeCycleManager = injector.getInstance(LifeCycleManager.class);
+        HttpServerInfo httpServerInfo = injector.getInstance(HttpServerInfo.class);
+        baseUri = httpServerInfo.getHttpUri();
+    }
+
+    private static <T> T proxyHandler(ThriftMetastore delegate, Class<T> iface)
+    {
+        return newProxy(iface, (proxy, method, args) -> switch (method.getName()) {
+            case "getAllDatabases" -> delegate.getAllDatabases();
+            case "getDatabase" -> {
+                Optional<Database> optionalDatabase = delegate.getDatabase(args[0].toString());
+                yield optionalDatabase.orElseThrow(() -> new NoSuchObjectException(""));
+            }
+            default -> throw new UnsupportedOperationException();
+        });
+    }
+
+    public int getPort()
+    {
+        return baseUri.getPort();
+    }
+
+    @Override
+    public void close()
+            throws IOException
+    {
+        lifeCycleManager.stop();
+    }
+
+    private static class TestingThriftHttpServlet
```

Comment:
If we update to the recently released libthrift 0.19.0, we should be able to use `TServlet` directly, since they switched to `jakarta.servlet`

Reply:
Thanks a lot @electrum for letting me know. I rebased and got the libthrift version to 0.19.0. The `THttpClient` in libthrift 0.19.0 takes a http5 client. I need to understand how I can change my code in `DefaultThriftMetastoreClientFactory` to instantiate a httpclient5. If you have any pointers for doing that I would appreciate it.

Reply:
I rebased and got the 0.19.0 upgrade which was done already. To support 0.19.0 libthrift I had to pull in http5 client jars because the `THttpClient` in 0.19.0 takes in a http5 client. Hope that is okay.